sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/TracerProvider.java

```diff
@@ -0,0 +1,155 @@
+                break;
+            default:
+                // ON_SUBSCRIBE and ON_NEXT don't have the information to end the span so just return.
+                break;
+        }
+    }
+
+    public <T extends CosmosResponse<?>> Mono<T> traceEnabledCosmosResponsePublisher(Mono<T> resultPublisher,
+                                                                                                             Context context,
+                                                                                                             String spanName,
+                                                                                                             String databaseId,
+                                                                                                             String endpoint) {
+        return traceEnabledPublisher(resultPublisher,  context, spanName,databaseId, endpoint,
+            (T response) -> response.getStatusCode());
+    }
+
+    public <T> Mono<CosmosItemResponse<T>> traceEnabledCosmosItemResponsePublisher(Mono<CosmosItemResponse<T>> resultPublisher,
+                                                                                   Context context,
+                                                                                   String spanName,
+                                                                                   String databaseId,
+                                                                                   String endpoint) {
+        return traceEnabledPublisher(resultPublisher, context, spanName,databaseId, endpoint,
+            CosmosItemResponse::getStatusCode);
+    }
+
+    public <T> Mono<T> traceEnabledPublisher(Mono<T> resultPublisher,
+                                             Context context,
+                                             String spanName,
+                                             String databaseId,
+                                             String endpoint,
+                                             Function<T, Integer> statusCodeFunc) {
+        final AtomicReference<Context> parentContext = new AtomicReference<>(Context.NONE);
+        Optional<Object> callDepth = context.getData(COSMOS_CALL_DEPTH);
+        final boolean isNestedCall = callDepth.isPresent();
+        return resultPublisher
+            .doOnSubscribe(ignoredValue -> {
+                if (isEnabled() && !isNestedCall) {
+                    parentContext.set(this.startSpan(spanName, databaseId, endpoint,
+                        context));
+                }
+            }).doOnSuccess(response -> {
+                if (isEnabled() && !isNestedCall) {
+                    this.endSpan(parentContext.get(), Signal.complete(), statusCodeFunc.apply(response));
+                }
+            }).doOnError(throwable -> {
+                if (isEnabled() && !isNestedCall) {
+                    this.endSpan(parentContext.get(), Signal.error(throwable), ERROR_CODE);
```

Comment:
Rather than this defaulting to 0 we should get the error status code from the Throwable itself.

Reply:
We are already doing at the end by converting throwable to cosmos [here  ](https://github.com/simplynaveen20/azure-sdk-for-java/blob/8af1cc9d56a405a9ec2bb72bac142b9d8c02890e/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/TracerProvider.java#L91). And if it is non cosmos error , this 0 will be used